### PR TITLE
magento-engcom/community-portal#77 Only users with privileges should see the edit and create project buttons

### PR DESCRIPTION
**Pull request by XiyaMagento**
magento-engcom/community-portal#77 Only users with privileges should see the edit and create project buttons

- [x] Enforce checks on related project endpoints for users to have write:project scope
- [x] Write tests to make sure that scope checking works correctly
- [x] Hide the create/edit button and enforce checks before showing the buttons

Testing
- Added two users, one with write:project scope and the other without write:project scope

Test cases
- Check that the user with write:project scope can create project and the other cannot
- Check that the user with write:project scope can edit project and the other cannot
- Check that the user with write:project scope can update project status and the other cannot

### CODE REVIEW

File: frontend/src/components/WithAuth.tsx [hunk 1]
```diff
@@ -31,31 +33,51 @@ interface WithAuthStateProps {
...
   class WithAuth extends React.Component<WithAuthProps & WithAuthStateProps & WithAuthDispatchProps, {}> {
     constructor(props: WithAuthProps) {
       super(props);
     }
     render() {
-      const { role } = this.props.user;
-      if (allowedRoles.includes(role)) {
-        return <WrappedComponent {...this.props} />;
+      const { role, scopes } = this.props.user;
+      console.log('WithAuth.tsx debugging - scopes:', scopes);
```
Comment (vrann) [inline frontend/src/components/WithAuth.tsx]: please remove debugging from the frontend

Comment (XiyaMagento) [reply frontend/src/components/WithAuth.tsx]: The debugging line has been removed

Comment (XiyaMagento) [reply frontend/src/components/WithAuth.tsx]: Summary of the test cases has been added to the issue description

File: frontend/src/components/WithAuth.tsx [hunk 2]
```diff
@@ -31,31 +33,51 @@ interface WithAuthStateProps {
...
+
+      // do not render if user has no required scopes
+      if (
+        compulsoryScopes !== undefined
```
Comment (athensbird) [inline frontend/src/components/WithAuth.tsx]: Looks like ```typeof compulsoryScopes !== 'undefined'``` would better for checking undeclared variables. See [here] (https://stackoverflow.com/questions/3390396/how-to-check-for-undefined-in-javascript) for more info

File: frontend/src/components/WithAuth.tsx [hunk 3]
```diff
@@ -31,31 +33,51 @@ interface WithAuthStateProps {
...
+        && Array.isArray(compulsoryScopes)
+        && compulsoryScopes.length > 0
+      ) {
+        if (
+          scopes === undefined
```
Comment (athensbird) [inline frontend/src/components/WithAuth.tsx]: Same issue for checking `undefined`

File: frontend/src/components/WithAuth.tsx [hunk 4]
```diff
@@ -31,31 +33,51 @@ interface WithAuthStateProps {
...
+        ) {
+          return null;
+        }
+      }
+
+      if (!allowedRoles.includes(role)) {
+        return <Login
+          clientId="668e0b6c450cc783f267"
+          scope=""
+          redirectUri="http://localhost:3030/auth"
```
Comment (athensbird) [inline frontend/src/components/WithAuth.tsx]: Are we still using the hard-coded credentials or passing it dynamically through local environment variable?

File: frontend/src/components/WithAuth.tsx (hunk 2 above)
Comment (XiyaMagento) [reply frontend/src/components/WithAuth.tsx]: Thanks! It has been changed accordingly

File: frontend/src/components/WithAuth.tsx (hunk 3 above)
Comment (XiyaMagento) [reply frontend/src/components/WithAuth.tsx]: Changed to typeof scopes === 'undefined'

File: frontend/src/components/WithAuth.tsx (hunk 4 above)
Comment (XiyaMagento) [reply frontend/src/components/WithAuth.tsx]: I think moving forward, we need to pass it in using local environment variable. I'm not sure if passing through local environment variables is working right now. But if so, I'll update this part to the newest version while merging back to sprint-dev

**Comment by XiyaMagento** [comment]
All comments addressed. Merging this pull request now.

File: frontend/src/components/WithAuth.tsx (hunk 4 above)
Comment (vrann) [reply frontend/src/components/WithAuth.tsx]: after webpack config branch is merged, this should be updated

Comment (XiyaMagento) [reply frontend/src/components/WithAuth.tsx]: Okay, noted

**Pull request merged**